.dockerhub/source.Dockerfile

```diff
@@ -0,0 +1,177 @@
+# syntax=docker/dockerfile:experimental
+
+# Use experimental buildkit for faster builds
+# https://github.com/moby/buildkit/blob/master/frontend/dockerfile/docs/experimental.md
+# Use `--progress=plain` to use plane stdout for docker build
+#
+# Example build command:
+# This determines which version of the ROS2 code base to pull
+# export ROS2_BRANCH=main
+# export DOCKER_BUILDKIT=1
+# docker build \
+#   --tag nav2:source \
+#   --file source.Dockerfile ../
+#
+# Omit the `--no-cache` if you know you don't need to break the cache.
+# We're only building on top of a ros2 devel image to get the basics
+# prerequisites installed such as the apt source, rosdep, etc. We don't want to
+# actually use any of the ros release packages. Instead we are going to build
+# everything from source in one big workspace.
+
+ARG FROM_IMAGE=osrf/ros2:devel
+ARG UNDERLAY_WS=/opt/underlay_ws
+ARG OVERLAY_WS=/opt/overlay_ws
+
+# multi-stage for caching
+FROM $FROM_IMAGE AS cacher
+
+# clone ros2 source
+ARG ROS2_BRANCH=master
+ARG ROS2_REPO=https://github.com/ros2/ros2.git
+WORKDIR $ROS2_WS/src
+RUN git clone $ROS2_REPO -b $ROS2_BRANCH && \
+    vcs import ./ < ros2/ros2.repos && \
+    find ./ -name ".git" | xargs rm -rf
+
+# clone underlay source
+ARG UNDERLAY_WS
+WORKDIR $UNDERLAY_WS/src
+COPY ./tools/underlay.repos ../
+RUN vcs import ./ < ../underlay.repos && \
+    find ./ -name ".git" | xargs rm -rf
+
+# copy overlay source
+ARG OVERLAY_WS
+WORKDIR $OVERLAY_WS/src
+COPY ./ ./ros-planning/navigation2
+RUN colcon list --names-only | cat > ../packages.txt
+
+# remove skiped packages
+WORKDIR /opt
+RUN find ./ \
+      -name "AMENT_IGNORE" -o \
+      -name "CATKIN_IGNORE" -o \
+      -name "COLCON_IGNORE" \
+      | xargs dirname | xargs rm -rf || true && \
+    colcon list --paths-only \
+      --packages-skip-up-to  \
+        $(cat $OVERLAY_WS/packages.txt | xargs) \
+      | xargs rm -rf
+
+# copy manifests for caching
+RUN mkdir -p /tmp/opt && \
```

Comment:
What's the aim for caching the package.xml files?

Reply:
In short this allows you to modify the local codebase in the repo without busting the docker build cache for cloning/building the underlay or re-installing package dependencies. These caches are selectively busted only if the .repos file or local package manifests are changed; and so this saves a load of dev time when build/testing locally using docker.

* https://github.com/ros-planning/navigation2/blob/a8df3f014ad9f1527086b79c1755cbb679897108/doc/continuous_integration/dockerfile.md#cacher-stage

This fancy multi staging is a workaround for the fact that docker's `COPY` directive does not have full globbing support, e.g. `COPY ./**/package.xml ./` does not work as expected. But still, more clever things can be done with multi staging.

* https://github.com/moby/moby/issues/39530#issuecomment-637905128

Reply:
Isn't that bad if we don't break the cache for the overlay workspace only on new releases or a new pkg? There's lots of changes between formal releases

Reply:
It's only for local dev builds, as builds on DockerHub are set to not use caching. If your building locally and haven't needed to change the package or repos file, and feel your docker build cache is getting stale, or find yourself with a speedy internet connection (some contributors don't have this at home), then you can use the `--no-cache` argument to docker build as mentioned in the Dockerfile to build from scratch.